agent/command/macsign.go

```diff
@@ -0,0 +1,209 @@
+func (macSign *macSign) Name() string { return "mac.sign" }
+
+// macSign-specific implementation of ParseParams.
+func (macSign *macSign) ParseParams(params map[string]interface{}) error {
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result: macSign,
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := decoder.Decode(params); err != nil {
+		return errors.Wrapf(err, "error decoding %s params", macSign.Name())
+	}
+
+	return macSign.validate()
+}
+
+func (macSign *macSign) validate() error {
+	catcher := grip.NewSimpleCatcher()
+
+	// make sure the command params are valid
+	if macSign.KeyId == "" {
+		catcher.Add(errors.New("key_id cannot be blank"))
+	}
+	if macSign.Secret == "" {
+		catcher.Add(errors.New("secret cannot be blank"))
+	}
+	if macSign.LocalZipFile == "" {
+		catcher.Add(errors.New("local_zip_file cannot be blank"))
+	}
+	if macSign.OutputZipFile == "" {
+		catcher.Add(errors.New("output_zip_file cannot be blank"))
+	}
+	if macSign.ServiceUrl == "" {
+		catcher.Add(errors.New("service_url cannot be blank"))
+	}
+	if macSign.Verify && runtime.GOOS != "darwin" {
+		catcher.Add(errors.New("verify is only supported in macOS"))
+	}
+	if !(macSign.ArtifactType == "" || macSign.ArtifactType == "binary" || macSign.ArtifactType == "app") {
+		catcher.Add((errors.New(`artifact_type needs to be either blank, "binary" or "app"`)))
+	}
+	if macSign.Notarize && macSign.BundleId == "" {
+		catcher.Add(errors.New("if notarization is requested, bundle_id cannot be blank"))
+	}
+
+	return catcher.Resolve()
+}
+
+// Apply the expansions from the relevant task config (including restricted expansion)
+// to all appropriate fields of the macSign.
+func (macSign *macSign) expandParams(conf *internal.TaskConfig) error {
+	var err error
+	if err = util.ExpandValues(macSign, conf.GetExpansionsWithRestricted()); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if filepath.IsAbs(macSign.LocalZipFile) {
+		macSign.workDir = ""
+	} else {
+		macSign.workDir = conf.WorkDir
+	}
+
+	return nil
+}
+
+func (macSign *macSign) shouldRunForVariant(buildVariantName string) bool {
+	//If no buildvariant filter, run for everything
+	if len(macSign.BuildVariants) == 0 {
+		return true
+	}
+
+	//Only run if the buildvariant specified appears in the list.
+	return utility.StringSliceContains(macSign.BuildVariants, buildVariantName)
+}
+
+// Implementation of Execute. Expands the parameters,
+// and then execute macOS signing and/or notarization.
+func (macSign *macSign) Execute(ctx context.Context,
+	comm client.Communicator, logger client.LoggerProducer, conf *internal.TaskConfig) error {
+
+	if !macSign.shouldRunForVariant(conf.BuildVariant.Name) {
+		logger.Task().Infof("Skipping macsign of local file %v for variant %v",
+			macSign.LocalZipFile, conf.BuildVariant.Name)
+		return nil
+	}
+
+	// expand necessary params
+	if err := macSign.expandParams(conf); err != nil {
```

Comment:
Should you expandParams before checking if you should skip the variant, just in case something needs to be populated?